# L1: Information Retrieval

In this lab you will apply basic techniques from information retrieval to implement the core of a minimalistic search engine. The data for this lab consists of a collection of app descriptions scraped from the [Google Play Store](https://play.google.com/store/apps?hl=en). From this collection, your search engine should retrieve those apps whose descriptions best match a given query under the vector space model.

## Data set

The app descriptions come in the form of a compressed [JSON](https://en.wikipedia.org/wiki/JSON) file. Start by loading this file into a [Pandas](https://pandas.pydata.org) [DataFrame](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe).

In [1]:
import bz2
import pandas as pd

with bz2.open("app-descriptions.json.bz2") as source:
    df = pd.read_json(source)

In Pandas, a DataFrame is a table with indexed rows and labelled columns of potentially different types. Data in a DataFrame can be accessed in various ways, including by row and by column. To give an example, the code in the next cell shows rows 200–204:

In [2]:
df[200:205]

name  \
200  Brick Breaker Star: Space King   
201      Brick Classic - Brick Game   
202     Bricks Breaker - Glow Balls   
203            Bricks Breaker Quest   
204             Brothers in Arms® 3   

                                           description  
200  Introducing the best Brick Breaker game that e...  
201  Classic Brick Game!\n\nBrick Classic is a popu...  
202  Bricks Breaker - Glow Balls is a addictive and...  
203  How to play\n- The ball flies to wherever you ...  
204  Fight brave soldiers from around the globe on ...

As you can see, there are two labelled columns: `name` (the name of the app) and `description` (a textual description). The code in the next cell shows how to acess fields from the description column.

## Problem 1: Preprocessing

Your first task is to implement a preprocessor for your search engine. In the vector space model, *preprocessing* refers to any kind of transformation that is applied before a text is vectorized. Here you can restrict yourself to a very simple preprocessing: tokenization, stop word removal, and lemmatization.

To implement your preprocessor, you can use [spaCy](https://spacy.io). Make sure that you read the [Linguistic annotations](https://spacy.io/usage/spacy-101#annotations) section of the spaCy&nbsp;101; that section contains all the information that you need for this problem (and more).

Implement your preprocessor by completing the skeleton code in the next cell, adding additional code as you feel necessary.

In [149]:
import spacy
def preprocess(text):
    # TODO: Replace the next line with your own code.
    
    # the model here needs to be install separately
    nlp = spacy.load("en_core_web_sm",
                     disable=["parser","ner","tagger"])        
    doc = nlp(text)
    mycode = []
    for token in doc:
#         print(token.text, token.lemma_, token.pos_, token.tag_, token.dep_,
#                     token.shape_, token.is_alpha, token.is_stop)
        if (not token.is_stop and token.lemma_.isalpha() ):
            mycode.append(token.lemma_)
    return mycode

Your implementation should conform to the following specification:

<strong>preprocess</strong> (<em>text</em>)

> Preprocesses given text by tokenizing it, removing any stop words, replacing each remaining token with its lemma (base form), and discarding all lemmas that contain non-alphabetical characters. Returns the list of remaining lemmas (represented as strings).

**Tip:** To speed up the preprocessing, you can disable loading those spaCy components that you do not need, such as the part-of-speech tagger, parser, and named entity recognizer. See [here](https://spacy.io/usage/processing-pipelines#disabling) for more information about this.

Test your implementation by running the following cell:

In [150]:
preprocess("Apple is looking at buying U.K. startup for $1 billion")

['Apple', 'look', 'buy', 'startup', 'billion']

This should give the following output:


## Problem 2: Vectorizing

Your next task is to vectorize the data – and more specifically, to map each app description to a tf–idf vector. For this you can use the [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) class from [scikit-learn](https://scikit-learn.org/stable/). Make sure to specify your preprocessor from the previous problem as the `tokenizer` &ndash; not the `preprocessor`! &ndash; for the vectorizer.

In [151]:
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer

# TODO: Replace the next line with your own code.
# X = np.zeros((len(df), 1))

vectorizer = TfidfVectorizer(use_idf=True, tokenizer=preprocess)
X = vectorizer.fit_transform(df["description"])


Test your implementation by running the following cell:

In [152]:
X.shape 
# with tags NNS NNPS 21457

(1614, 20651)

This should show the dimensions of the matrix `X` to be 1614 × 20651.

## Problem 3: Finding terms with low/high idf

Recall that the inverse document frequency (idf) of a term is the lower the more documents from a given collection the term appears in. To get a better understanding for this concept, your next task is to write code to find out which terms have the lowest/highest idf with respect to the app descriptions.

Start by sorting the terms in increasing order of idf, breaking ties by falling back on alphabetic order.

In [153]:
# # TODO: Replace the next line with your own code.
# terms = []

idfs = vectorizer.idf_
tuples = zip(vectorizer.get_feature_names(), idfs)
terms = sorted(tuples, key=lambda x:(x[1], x[0]), reverse=False)

Now, print out the 10 terms at the top/end of the sorted list:

In [154]:
terms[:10]

[('game', 1.2961804135767734),
 ('play', 1.4126113246317116),
 ('feature', 1.4723593429381947),
 ('free', 1.5846954723324462),
 ('new', 1.6656645348661132),
 ('world', 1.781792314708555),
 ('time', 1.7885812070422413),
 ('app', 1.8388711328943845),
 ('fun', 1.8591323180342063),
 ('well', 1.872377544784227)]

## Problem 4: Retrieving

To complete the search engine, your last task is to write a function that returns the most relevant app descriptions for a given query. An easy way to do solve this task is to use scikit-learn&rsquo;s [NearestNeighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) class. That class implements unsupervised nearest neighbours learning, and allows you to easily find a predefined number of app descriptions whose vector representations are closest in distance to the query vector.

In [155]:
from sklearn.neighbors import NearestNeighbors
def search(query):
    # TODO: Replace the next line with your own code.
    neigh = NearestNeighbors(metric="cosine")
    neigh.fit(X)
    q = vectorizer.transform([query])

    idx = neigh.kneighbors(q, 10, return_distance=False)
    return df.iloc[idx[0]]

Your implementation should conform to the following specification:

<strong>search</strong> (<em>query</em>)

> Returns the 10 app descriptions most similar (in terms of cosine similarity) to the given query as a Pandas DataFrame.

Test your implementation by running the following cell:

In [156]:
search("dodge trains")

name  \
1301                             Subway Surfers   
1300                     Subway Princess Runner   
1428                      Train Conductor World   
998              No Humanity - The Hardest Game   
1394                                 Tiny Rails   
1429        Train for Animals - BabyMagica free   
1168                                       Rush   
1077                Polar Flow – Sync & Analyze   
1286  Strava: Track Running, Cycling & Swimming   
1465            Virus War - Space Shooting Game   

                                            description  
1301  DASH as fast as you can! \nDODGE the oncoming ...  
1300  Subway princess runner, Bus run, forest rush w...  
1428  Master and manage the chaos of international r...  
998   2M+ Downloads All Over The World!\n\n* IGN Nom...  
1394  All aboard for an adventure around the world!\...  
1429  🚂 BabyMagica "Train for Animals" is a educatio...  
1168  Are you ready for a thrilling ride?\n\nRush th...  
1077  Polar Flow is a sports, fitness, and activity ...  
1286  Track your fitness with Strava activity tracke...  
1465  Warning! Virus invasion! Destroy them with you...

The top hit in the list should be *Subway Surfers*.

## Problem 5: Keyword extraction

A simple method for extracting salient keywords from a document is to pick the $k$ terms with the highest tf–idf value. Your last task in this lab is to implement this method. More specifically, we ask you to implement a function `keywords` that extracts keywords from a given text.

In [233]:
def keywords(text, n=10):
    # TODO: Replace the next line with your own code.
    a = vectorizer.transform([df["description"][1428]])
    vol = vectorizer.get_feature_names() 
    idx = np.argsort(a.toarray()).tolist()[0][-n:][::-1]      # idx is the indices in the whole terms list
    return [vol[i] for i in idx]                             # np.flipud can do reserve as well

Your implementation should conform to the following specification:

<strong>keywords</strong> (<em>text</em>, <em>n</em> = 10)

> Returns a list with the $n$ (default value: 10) most salient keywords from the specified text, as measured by their tf–idf value relative to the collection of app descriptions.

Test your implementation by running the following cell:

In [234]:
print(keywords(df["description"][1428]))

['train', 'railroad', 'railway', 'rail', 'chaos', 'crash', 'tram', 'haul', 'overcast', 'railyard']


This should give the following output:

<div class="alert alert-info">
    Please read the General information section on the lab web page before submitting this notebook!
</div>